### ISSUE TITLE
cAdvisor e2e failing 100% on core OS

### ISSUE THREAD

**Issue by timstclair** [google/cadvisor#1344]
Error:
```
F0620 11:09:44.381011   18340 runner.go:290] Error 0: error on host e2e-cadvisor-coreos-beta: command "godep" ["go" "test" "--timeout" "15m0s" "github.com/google/cadvisor/integration/tests/..." "--host" "e2e-cadvisor-coreos-beta" "--port" "8080" "--ssh-options" "-i /home/jenkins/.ssh/google_compute_engine -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o CheckHostIP=no -o StrictHostKeyChecking=no"] failed with error: exit status 1 and output: godep: WARNING: Godep workspaces (./Godeps/_workspace) are deprecated and support for them will be removed when go1.8 is released.
godep: WARNING: Go version (go1.6) & $GO15VENDOREXPERIMENT= wants to enable the vendor experiment, but disabling because a Godep workspace (Godeps/_workspace) exists
--- FAIL: TestDockerContainerSpec (0.98s)
	assertions.go:150: 
                        
	Location:	docker_test.go:225
		
	Error:		Not equal: "0" (expected)
			        != "" (actual)
		
	Messages:	Cpu mask should be "0", but is ""
		
FAIL
FAIL	github.com/google/cadvisor/integration/tests/api	29.639s
ok  	github.com/google/cadvisor/integration/tests/healthz	0.012s
godep: go exit status 1
```
[failing line](https://github.com/google/cadvisor/blob/master/integration/tests/api/docker_test.go#L225)

This has been failing 100% on e2e-cadvisor-coreos-beta since https://github.com/google/cadvisor/pull/1333 was merged.

@Random-Liu @pwittrock @euank

**Comment by timstclair** [comment]
@pwittrock do you know when you restarted the jenkins CI VMs? At first glance this doesn't look like the PR it started failing at would have caused it.

**Comment by timstclair** [comment]
Failing since kubekins build [#4096](http://kubekins.dls.corp.google.com/job/cadvisor-gce-e2e-ci/4096/) (sorry, Google internal only)

Actually, ignore my comment about https://github.com/google/cadvisor/pull/1333 - many builds succeeded on that revision prior to the failures starting.

I'm guessing the core-os VM auto-updated and is no longer setting the CPU mask. I'll ping the pr-builder jenkins to see if it's having the same issue.

**Comment by vishh** [comment]
@timstclair Were able to root cause this issue?

**Comment by timstclair** [comment]
No, I haven't tracked it down yet. It looks like this has started affecting the build jobs as well (it didn't yesterday). I think this lends credence to it being caused by a core OS update.

**Comment by timstclair** [comment]
Let me see if I can reproduce on a new GCE coreos instance.

**Comment by vishh** [comment]
We can disable CoreOS node from the e2e suite until this issue is resolved.
WDYT?

**Comment by timstclair** [comment]
I'll disable it for the builder.

**Comment by euank** [comment]
Do you have more information on how these images are setup and so on? For reference, node_e2e disable updates now with [this](https://github.com/kubernetes/kubernetes/blob/38182e91c93b8c0cce6c6aa3bcb0a39b1296b76e/test/e2e_node/environment/setup_host.sh#L45).

If the cadvisor tests fail on newer CoreOSs post-update, that sounds like something to triage and fix..

**Comment by timstclair** [comment]
Ok, I was able to reproduce this on a fresh coreOS beta image (`VERSION_ID=1068.2.0`). I just did:

1. Install go to home directory
2. go get godep & cadvisor
3. I had trouble building cadvisor, so I copied the binary built on another machine
4. run `build/integration.sh`, which surfaced the same error:
```
$ build/integration.sh 
>> starting cAdvisor locally
Waiting for cAdvisor to start ...
>> running integration tests against local cAdvisor
godep: WARNING: Godep workspaces (./Godeps/_workspace) are deprecated and support for them will be removed when go1.8 is released.
godep: WARNING: Go version (go1.6) & $GO15VENDOREXPERIMENT= wants to enable the vendor experiment, but disabling because a Godep workspace (Godeps/_workspace) exists
--- FAIL: TestDockerContainerSpec (0.51s)
        Location:       docker_test.go:225
	Error:  	Not equal: "0" (expected)
			        != "" (actual)
	Messages:	Cpu mask should be "0", but is ""
		
FAIL
FAIL	github.com/google/cadvisor/integration/tests/api	29.321s
ok  	github.com/google/cadvisor/integration/tests/healthz	0.004s
godep: go exit status 1
Integration tests failed
>> stopping cAdvisor
build/integration.sh: line 45:  5341 Killed                  sudo ./cadvisor --docker_env_metadata_whitelist=TEST_VAR
```

@euank can you take it from here?

**Comment by euank** [comment]
I'll take a guess that this is related to systemd >= 226's change in cgroup hierarchy, but not sure yet.

I was able to reproduce on a machine with systemd 226 and docker 1.11 launched with `--exec-opt native.cgroupdriver=systemd`.

The machine in question is gentoo, but I expect it'll reproduce broadly in that configuration. I'll dig further...

**Comment by euank** [comment]
It finds a cpuset root at both `/sys/fs/cgroup/cpuset/system.slice/docker-de6460b5039fa64f505cf383c15dc96d515bbc507c76ec0f7c06a00a5115002f.scope` and `/sys/fs/cgroup/cpuset/init.scope/system.slice/docker-de6460b5039fa64f505cf383c15dc96d515bbc507c76ec0f7c06a00a5115002f.scope` (note the `init.scope` bit), then tries to use the latter and doesn't find the files it expects and returns "". The first one would be correct. I think this is https://github.com/docker/docker/issues/16256#issuecomment-218935000 .. I'm giving a quick go at patching this similarly to as suggested to be sure.

**Comment by euank** [comment]
Upstream bug to point to as well: https://github.com/opencontainers/runc/issues/931

Our options are, I think:

1) Wait for runc + docker-on-coreos to be fixed (we might backport the fix)
2) Switch the docker test setup to use cgroupfs driver (non-default)
3) Switch to CoreOS stable where it's still using systemd 225 and thus should be okay
4) modify cadvisor to handle runc's error there and do the "right thing" when there are multiple cgroup paths for one container.

My preference is 3 to put off having to get a better solution, and hope that 1 happens in the meanwhile. Sound reasonable?

**Comment by timstclair** [comment]
(3) sounds reasonable to me, and I think we should do it anyway (filed https://github.com/google/cadvisor/issues/1361). Once the jenkins jobs are updated to use the [cAdvisor jenkins script](https://github.com/google/cadvisor/blob/master/build/jenkins_e2e.sh) (https://github.com/kubernetes/test-infra/pull/248) I'll add a coreos-stable VM.

**Comment by euank** [comment]
We should be able to update the CoreOS node if no one has already now that we've switched coreos to use cgroupfs by default.

I don't have access to the images referenced by `./build/jenkins_e2e.sh` to update appropriately, but any version of CoreOS right now should include that change. If it's possible to use unmodified coreos images as I updated the node e2e to do, that might also help.

**Comment by timstclair** [comment]
I'm not sure what (if anything) needs to be changed from the unmodified coreos image, so it might just work. If you're up for trying it and figuring out what (if anything) needs to be added, I'd certainly welcome the help :)

You can see the command used to run the tests [here](https://github.com/google/cadvisor/blob/master/build/jenkins_e2e.sh).

**Comment by dchen1107** [comment]
@euank I am assigning this one to you for delegating. We need better support for coreos as one of basic images for us. Re-assign it back to us or ask for help if you need. Thanks!

**Comment by euank** [comment]
I don't expect we'll need to do more than is done for the node_e2e stuff (user-data of https://github.com/euank/kubernetes/blob/5a5ba51b24c9e62aa775de1f568d365c2761aeb5/test/e2e_node/jenkins/coreos-init.json basically).

I'm on vacation for the next couple weeks, so I won't be able to verify that's true, and regardless someone with access to the jenkins account where these test instances run will need to start one up, unless we switch to a node_e2e type model where instances are launched and specified as part of code in this repository, not totally out of band.

cc @yifan-gu @crawford

**Comment by crawford** [comment]
Minor nit, the correct extension for that file should be `.ign` or `.ignition` per the [IANA MIME registration](http://www.iana.org/assignments/media-types/application/vnd.coreos.ignition+json).

**Comment by yifan-gu** [comment]
subscribe

**Comment by dims** [comment]
Closing this issue as it has been open since before January 1, 2024. If this is still relevant, please feel free to reopen or create a new issue.
